Dockerfile

```diff
@@ -1,74 +1,103 @@
-    case ${PRODUCT_NAME} in \
-        "ideaIC"|"ideaIU") \
-            wget -qO- https://download.jetbrains.com/idea/${PRODUCT_NAME}-${PRODUCT_VERSION}.tar.gz | tar -zxv --strip-components=1 -C /${PRODUCT_NAME}-${PRODUCT_VERSION} && \
-            ln -s /${PRODUCT_NAME}-${PRODUCT_VERSION}/bin/idea.sh /opt/run-ide.sh \
-            ;; \
-        "WebStorm") \
-            wget -qO- https://download.jetbrains.com/webstorm/${PRODUCT_NAME}-${PRODUCT_VERSION}.tar.gz | tar -zxv --strip-components=1 -C /${PRODUCT_NAME}-${PRODUCT_VERSION} && \
-            ln -s /${PRODUCT_NAME}-${PRODUCT_VERSION}/bin/webstorm.sh /opt/run-ide.sh \
-            ;; \
-    esac && \
-    mkdir -p ${BASE_MOUNT_FOLDER}/${PRODUCT_NAME} && \
-    mkdir /etc/default/jetbrains && \
-    for f in "${BASE_MOUNT_FOLDER}" "/${PRODUCT_NAME}-${PRODUCT_VERSION}" "/etc/passwd" "/etc/default/jetbrains"; do \
-      echo "Changing permissions on ${f}" && chgrp -R 0 ${f} && \
-      chmod -R g+rwX ${f}; \
-    done && \
-    echo "idea.config.path=${BASE_MOUNT_FOLDER}/${PRODUCT_NAME}/config" > /${PRODUCT_NAME}-${PRODUCT_VERSION}/bin/idea.properties && \
-    echo "idea.system.path=${BASE_MOUNT_FOLDER}/${PRODUCT_NAME}/caches" >> /${PRODUCT_NAME}-${PRODUCT_VERSION}/bin/idea.properties && \
-    echo "idea.plugins.path=${BASE_MOUNT_FOLDER}/${PRODUCT_NAME}/plugins" >> /${PRODUCT_NAME}-${PRODUCT_VERSION}/bin/idea.properties && \
-    echo "idea.log.path=${BASE_MOUNT_FOLDER}/${PRODUCT_NAME}/logs" >> /${PRODUCT_NAME}-${PRODUCT_VERSION}/bin/idea.properties
+#
+# Copyright 2019-2020 JetBrains s.r.o.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
 
-# Copy fluxbox configuration
-COPY --chown=0:0 config/fluxbox /home/user/.fluxbox/init
 
-# Copy predefined configs
-COPY --chown=0:0 config/etc /etc/
+# Stage 1. Prepare JetBrains IDE with Projector.
+#   1. Downloads JetBrains IDE packaging by given downloadUrl build argument.
+#   2. If buildGradle build argument is set to false, then consumes built Projector assembly from the host.
+#       2.1 Otherwise starts Gradle build of Projector Server and Projector Client.
+#   3. Copies static files to the Projector assembly (entrypoint, launcher, configuration).
+FROM registry.access.redhat.com/ubi8-minimal:8.3-298 as projectorAssembly
+ENV PROJECTOR_DIR /projector
+ENV JAVA_HOME /usr/lib/jvm/java-11
+ARG downloadUrl
+ARG buildGradle
+ADD jetbrains-editor-images/static $PROJECTOR_DIR
+ADD projector-client $PROJECTOR_DIR/projector-client
```

Comment:
might be worth adding a comment here that these sources require that you first run https://github.com/che-incubator/jetbrains-editor-images/blob/projector/clone-projector.sh so as to get more sources into the repo.

productizing this downstream will require that the sources be prefetched *and committed* as we can't:
* run arbitrary shell scripts as part of the docker build
* connect to the internet during the docker build

Reply:
Yep, make sense. Will add the comment about it.

Reply:
other thing to note here is that we're currently depending on your private fork for the https://github.com/vzhukovs/projector-client.git sources... which is generally not advised for either Che projects or downstream product. 

Assuming we will need to maintain this fork, can we move it to the eclipse-che organization or the che-incubator one, so it's at least an "official" repo?

Reply:
Added comment in Dockerfile: https://github.com/che-incubator/jetbrains-editor-images/blob/projector/Dockerfile#L25-L31

Reply:
As for forked repository. I've added an ability to clone upstream repositories instead of forks and patching them with `.patch` files: https://github.com/che-incubator/jetbrains-editor-images/tree/projector/patches This will allow us not to stick to forks. But as for Projector Client, I've sticked to the specific commit, because it releases seldom. cc @benoitf